contributors/devel/sig-testing/e2e-tests.md

```diff
@@ -587,9 +595,67 @@ refine requirements:
+    `[Feature:.+]`, but for historic reasons and the sake of brevity they don't
+    have that prefix when embedded in test names. They *do* have that prefix in the
+    Ginkgo v2 label, so use e.g. `--filter-label=Feature:Alpha`.
+
 Every test should be owned by a [SIG](/sig-list.md),
 and have a corresponding `[sig-<name>]` label.
 
+## Selecting tests to run
+
+See https://onsi.github.io/ginkgo/#filtering-specs for a general introduction.
+
+Focusing on a specific test by its name is useful when interactively running
+just one or a few related tests. The test name is a concatenation of multiple
+strings. To get a list of all full test names, run:
+
+```console
+$ e2e.test -list-tests
+The following spec names can be used with 'ginkgo run --focus/skip':
+    test/e2e/apimachinery/watchlist.go:41: [sig-api-machinery] API Streaming (aka. WatchList) [Serial] [Feature:WatchList] should be requested when ENABLE_CLIENT_GO_WATCH_LIST_ALPHA is set
+    test/e2e/apimachinery/flowcontrol.go:65: [sig-api-machinery] API priority and fairness should ensure that requests can be classified by adding FlowSchema and PriorityLevelConfiguration
+    test/e2e/apimachinery/flowcontrol.go:190: [sig-api-machinery] API priority and fairness should ensure that requests can't be drowned out (fairness)
+...
+```
+
+Or within the Kubernetes repo:
+
+```console
+$ go test -v ./test/e2e -args -list-tests
+The following spec names can be used with 'ginkgo run --focus/skip':
+    test/e2e/apimachinery/watchlist.go:41: [sig-api-machinery] API Streaming (aka. WatchList) [Serial] [Feature:WatchList] should be requested when ENABLE_CLIENT_GO_WATCH_LIST_ALPHA is set
+...
+```
+
+The same works for other Kubernetes E2E suites, like `e2e_node`.
+
+In Prow jobs, selection by labels is often simpler. See
+[below]((#kinds-of-tests) for documentation of the different labels that are in
+use. A full list of labels used by a specific E2E suite can be obtained with
+`--list-labels`.
+
+A common pattern is to run only tests which have no special cluster setup
+requirements:
+
+    --filter-label=!/^Feature:/
+
+Feature owners have to ensure that tests excluded that way from shared CI
+jobs are executed in dedicated jobs (more on CI below):
+
+    --filter-label=Feature:MyAwesomeFeature
+
+As a special case, excluding tests with special requirements *except* for some
+particular feature is possible with a carefully crafted regular expression that
+[emulates a negative lookahead (`Feature:(?!Alpha)`)](https://stackoverflow.com/a/76064059/222305).
+Negative lookahead is not supported by Go regexp and thus ginkgo. The following runs all
+tests that have no special requirements and those which have
+`Feature:Alpha`. It works because any word after `Feature:` other than exactly `Alpha`
+will match one of the alternatives, but `Alpha` itself doesn't:
+
+    --filter-label=!/^Feature:([^A]|A[^l]|Al[^p]|Alp[^h]|Alph[^a]|Alpha.)/
```

Comment:
The alternative is to make Ginkgo use a regexp implementation which supports negative lookahead. But that then adds a new dependency to Kubernetes. I don't see a big need for this kind of special "exclude features except for these", so covering those few cases with this special regexp is the lesser evil?

Reply:
I would rather prefer to avoid complex expressions

Reply:
So do I, but I can't think of a simpler solution. Using `Graduation:Alpha` or `Stability:Beta` doesn't work (try writing a `--label-filter` for `ci-kubernetes-e2e-kind-alpha-features` based on that and you'll see why).

Reply:
we could provide our own command-line syntax and have `e2e.test` convert it into an ugly regex for ginkgo's benefit?

Reply:
This enters a grey area: it assumes that the `-focus/skip/label-filter` parameters that get passed to the `ginkgo` CLI command are just getting passed through to the e2e.test binary and all the parsing and logic happens there. This happens to work and we already rely on it in various places by passing `-ginkgo.focus/skip` *only* to the e2e.test binary, so I suppose we can assume that this works.

But what would such a command-line syntax look like? How does it interact with the other parameters? Is this really worth adding a special case for?

Reply:
... or we can work with @onsi on defining  much richer semantics on test filtering ...

Reply:
Mirroring what I discussed with @aojea on Slack: coming up with something that avoids the negative lookahead workaround and/or regexp entirely is something that can be added later. For now we could proceed as proposed here.

/cc @BenTheElder

Reply:
> But what would such a command-line syntax look like?

I was thinking mostly about the case of "I want to skip all features except for X, Y, and Z", which _might_ be common enough to merit special syntax?

But anyway, yeah, this is "later" stuff